src/downloader.rs

```diff
@@ -7,18 +8,50 @@ use crate::warn;
 pub struct Downloader {
     client: reqwest::blocking::Client,
     tries: usize,
+    auth_map: HashMap<Option<String>, (String, Option<String>)>,
+}
+
+///Parse HTTP authentication credentials from string iterable
+fn parse_auth(auth: &[String]) -> Option<(String, Option<String>, Option<String>)> {
+    // Convert any empty strings to None
+    let auth: Vec<Option<String>> = auth
+        .iter()
+        .map(|s| match s.as_ref() {
+            "" => None,
+            s => Some(s.to_string()),
+        })
+        .collect();
+    match auth.as_slice() {
+        [] => None,
+        [None, ..] => None, // Empty username should be considered no-op
```

Comment:
I think that would make sense! Any preference on how errors are handled here whether it's a custom error type or something else?

Reply:
Atm, SuckIT does a poor job in term of error handling. Return an ``Invalid argument`` using ``Error`` here or consider adding a check function [here](https://github.com/Skallwar/suckit/blob/794d8ddd533097019f02a67d99bcca2378b43e0d/src/args.rs#L111)

Reply:
Sounds good!

Reply:
It looks like it would be more straightforward to use an error than a check function because of the additional logic we're using around pulling the host information, but to clarify is there a specific `structopt` `Error` struct you're referring to or should `parse_auth` return a `Result` and call `panic!` if it fails?

Reply:
This should be updated in the latest commit which just triggers an error with `unwrap` if `parse_auth` is an `Err`, but let me know if I misunderstood

Reply:
> but to clarify is there a specific structopt Error struct you're referring to

No, there is none.

> should parse_auth return a Result and call panic! if it fails?

Panic is not a great solution. Maybe something like what has been done [here](https://github.com/Skallwar/suckit/blob/794d8ddd533097019f02a67d99bcca2378b43e0d/src/args.rs#L87) and [here](https://github.com/Skallwar/suckit/blob/794d8ddd533097019f02a67d99bcca2378b43e0d/src/args.rs#L118) could be cool. This way, structop will handle the error by itself with some pretty printing with a lot more context than an panic in the downloader. I like the idea of the argument being checked in one place and the rest of the program running with full confidence in his arguments

Reply:
Thanks for clarifying, the issue with that is we need access to the `origin` value in determining whether the values are valid because the host will default to the origin's host value, and `origin.host_str` returns an `Option` so I'm not sure we'd be able to do it in the same way

Reply:
Have you tried it?

Reply:
I didn't see any way of accessing multiple arg values in a `parse` function going through the structopt docs, but if I missed something I'm happy to give it a try